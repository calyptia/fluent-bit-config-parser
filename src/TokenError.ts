const formatError = (message: string, filePath: string, line: number, col: number): string =>
  `${filePath}: ${line}:${col} ${message}`;
export class TokenError extends Error {
  line: number;
  col: number;
  message: string;

  filePath: string;
  constructor(message: string, filePath: string, line: number, col: number) {
    super(formatError(message, filePath, line, col));
    this.name = 'TokenError';
    this.message = message;
    this.filePath = filePath;
    this.line = line;
    this.col = col;
  }
}
