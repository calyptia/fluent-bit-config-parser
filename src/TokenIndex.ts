import type { FluentBitToken } from './constants';

export class TokenIndex {
  private map: Map<string, FluentBitToken[]>;
  constructor() {
    this.map = new Map();
  }
  clear() {
    this.map = new Map();
  }
  get(tokenId: string) {
    return this.map.get(tokenId);
  }
  set(tokenId: string, value: FluentBitToken) {
    const tokens = this.map.get(tokenId);
    if (tokens) {
      tokens.push(value);
    } else {
      this.map.set(tokenId, [value]);
    }

    return this;
  }
}
