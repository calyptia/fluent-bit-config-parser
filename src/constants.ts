import type { Token } from 'moo';
import { getBorderCharacters } from 'table';

export enum COMMANDS {
  OUTPUT = 'OUTPUT',
  INPUT = 'INPUT',
  FILTER = 'FILTER',
  SERVICE = 'SERVICE',
  PARSER = 'PARSER',
  CUSTOM = 'CUSTOM',
}
export enum TOKEN_TYPES_DIRECTIVES {
  SET = 'SET',
  INCLUDE = 'INCLUDE',
}
export enum TOKEN_TYPES {
  PROPERTIES = 'PROPERTIES',
  CLOSE_BLOCK = 'CLOSE_BLOCK',
  OPEN_BLOCK = 'OPEN_BLOCK',
  COMMAND = 'COMMAND',
  SPACE = 'SPACE',
  COMMENT = 'COMMENT',

  DIRECTIVES = 'DIRECTIVES',
}
export type FluentBitSection = {
  id: string;
  name?: string;
  command: COMMANDS;
  optional?: { [key: string]: unknown };
};

export interface FluentBitSchemaType extends FluentBitSection {
  __filePath: string;
}

export type FluentBitToken = Token & { filePath: string };
export const FLUENTBIT_REGEX = /(?<![#][ ]*)\[[A-Z]{1,}\]/g;

/** It will match @includes files as a valid Fluent Bit configuration.
 *  [Follow this link for an example](https://regex101.com/r/zrSRR2/1)
 */
export const FLUENTBIT_INCLUDE_REGEX = /(@include+\s.*){1,}/g;

export const EXCLUDED_TAGS = new Set(['service', 'parser', 'node', 'upstream']);

export const NO_STYLES_IN_TABLE = {
  border: getBorderCharacters('void'),
  columnDefault: {
    paddingLeft: 0,
    paddingRight: 1,
  },
  drawHorizontalLine: (): boolean => false,
};

export enum CUSTOM_SECTIONS_NAMES {
  FLUENTBIT_METRICS = 'fluentbit_metrics',
  CALYPTIA = 'calyptia',
}
