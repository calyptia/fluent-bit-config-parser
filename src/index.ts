export * from './parser';
export * from './schemaToString';
