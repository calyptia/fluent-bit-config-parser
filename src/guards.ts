import { FLUENTBIT_REGEX, COMMANDS, type FluentBitSchemaType, EXCLUDED_TAGS } from './constants';

export const isFluentBit = (config: string) => !!config.match(FLUENTBIT_REGEX);

export const isValidFluentBitSchemaType = (schema?: FluentBitSchemaType | null): schema is FluentBitSchemaType =>
  !!schema && !EXCLUDED_TAGS.has(schema.command.toLowerCase());

export const isString = (value?: string): value is string => typeof value === 'string';

export const isUsefulValidToken = (type?: string): boolean => isString(type) && !['space', 'comment'].includes(type);

export const isCommandType = (type?: string): type is COMMANDS =>
  isString(type) && Object.keys(COMMANDS).includes(type);

export const isCustomBlock = (block: FluentBitSchemaType) =>
  !!block.name && (block.name.includes('fluentbit_metrics') || block.name.includes('calyptia'));
