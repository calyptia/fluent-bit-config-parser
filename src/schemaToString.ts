import { table } from 'table';
import { NO_STYLES_IN_TABLE } from './constants';

import type { FluentBitSchemaType } from './constants';

const PROP_DEFAULT_INDENT = 3;

const getIndent = (indent: number): string => Array.from({ length: indent }).join(' ');
export function schemaToString(configBlocks: FluentBitSchemaType[], { propIndent = PROP_DEFAULT_INDENT }): string {
  const data = [] as string[][];
  const spanningCells = [];

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  for (const { command, id, optional, ...rest } of configBlocks) {
    data.push([`\n[${command}]`, '']);
    spanningCells.push({ col: 0, row: 3, colSpan: 3 });

    for (const [key, value] of Object.entries({ ...rest, ...optional })) {
      data.push([`${getIndent(propIndent)}${key}`, String(value)]);
      spanningCells.push({ col: 0, row: 0, colSpan: 0 });
    }
  }

  return table(data, NO_STYLES_IN_TABLE);
}
