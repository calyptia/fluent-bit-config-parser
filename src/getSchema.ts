import type { FluentBitSchemaType } from './constants';
import { isCustomBlock, isValidFluentBitSchemaType } from './guards';
export const getFluentBitSchema = (ast: FluentBitSchemaType[]): FluentBitSchemaType[] => {
  const schema = ast.filter((block) => isValidFluentBitSchemaType(block) && !isCustomBlock(block));
  return schema.map((configNode) => {
    const { name, id, command, optional } = configNode;
    return { command, name, optional, id };
  });
};
