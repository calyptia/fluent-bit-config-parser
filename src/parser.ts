import { v4 as uuidv4 } from 'uuid';
import { getFluentBitSchema } from './getSchema';
import { COMMANDS, type FluentBitSchemaType } from './constants';
import { isCommandType, isFluentBit, isUsefulToken } from './guards';
import { keywords, states, Token } from 'moo';
import { schemaToString } from './schemaToString';
import { readFileSync, realpathSync } from 'fs';
import { join } from 'path';
import { dirname, isAbsolute } from 'path/posix';
import { TokenError } from './TokenError';
function normalizeField(field: string) {
  const normalizedField = field.toLowerCase();
  return normalizedField === 'match_regex' ? 'match' : normalizedField;
}
enum TOKEN_TYPES {
  properties = 'PROPERTIES',
  closeBlock = 'CLOSE_BLOCK',
  openBlock = 'OPEN_BLOCK',
  command = 'COMMAND',
  include = 'INCLUDE',
}

type FLUENT_BIT_TOKEN = Token & { filePath: string };

const stateSet = {
  main: {
    [TOKEN_TYPES.openBlock]: { match: '[', push: 'block' },
    [TOKEN_TYPES.include]: { match: /@include+\s.*/, lineBreak: true },
    [TOKEN_TYPES.properties]: [
      {
        match: /\w+[-.*\d\w]+\s.*/,
        value: (value: string) => value.replace(/\s+/, ' ').trim(),
        lineBreaks: true,
      },
    ],
    space: { match: /\s+/, lineBreaks: true },
    comment: { match: /#.*/, lineBreaks: true },
  },
  block: {
    [TOKEN_TYPES.command]: {
      match: /\w+/,
      type: keywords(COMMANDS),
    },
    comment: { match: /#.*/, lineBreaks: true },
    [TOKEN_TYPES.closeBlock]: { match: ']', push: 'main' },
  },
};
export function tokenize(config: string, filePath: string): FLUENT_BIT_TOKEN[] {
  if (!config.replace(/\s/g, '')) {
    throw new TokenError('File is empty', filePath, 0, 0);
  }

  if (!isFluentBit(config)) {
    throw new TokenError('This file is not a valid Fluent Bit config file', filePath, 0, 0);
  }

  let tokens = [] as FLUENT_BIT_TOKEN[];
  const lexer = states(stateSet).reset(config);

  // We will expand every include first, looking for any missing paths and invalid tokens
  // https://github.com/calyptia/fluent-bit-config-parser/issues/15
  for (const token of lexer) {
    if (token.type === TOKEN_TYPES.include) {
      const [, includeFilePath] = token.value.split(' ');

      const fullPath = realpathSync(join(dirname(filePath), includeFilePath));

      let includeConfig = '';

      try {
        includeConfig = readFileSync(fullPath, { encoding: 'utf-8' });
      } catch (e) {
        throw new TokenError('Can not read file', fullPath, 0, 0);
      }

      const includeTokens = tokenize(includeConfig, filePath);
      tokens = [...tokens, ...includeTokens];
    } else {
      tokens.push({ ...token, filePath });
    }
  }

  return tokens;
}
export function tokensToAST(tokens: Token[]): FluentBitSchemaType[] {
  const configBlocks = [] as FluentBitSchemaType[];
  let block = {} as FluentBitSchemaType;
  let command = undefined as COMMANDS | undefined;

  for (const token of tokens) {
    // Ignoring spaces and comments
    if (!isUsefulToken(token.type)) {
      continue;
    }

    // If we find a valid command we begin collecting properties.
    if (isCommandType(token.type)) {
      command = token.value as COMMANDS;
      block = { command, id: uuidv4(), optional: {} };
      continue;
    }

    if (command) {
      if (token.type === TOKEN_TYPES.properties) {
        const [key, ...value] = token.value.split(' ');
        const attrName = normalizeField(key);
        const attrValue = value.join(' ');

        if (attrName === 'name') {
          block = { ...block, [attrName]: attrValue };
        } else {
          block = {
            ...block,
            optional: { ...block.optional, [attrName]: attrValue },
          };
        }

        continue;
      }

      if (token.type === TOKEN_TYPES.openBlock) {
        configBlocks.push({ ...block });
        block = {} as FluentBitSchemaType;
        command = undefined;
        continue;
      }
    }
  }

  return [...configBlocks, block];
}

function getFullPath(filePath: string) {
  return isAbsolute(filePath) ? filePath : realpathSync(filePath);
}

export class FluentBitSchema {
  private _filePath: string;
  private _source: string;
  private _tokens: Token[];
  constructor(source: string, filePath: string) {
    this._source = source;
    this._filePath = filePath;

    this._tokens = tokenize(source, getFullPath(filePath));
  }
  static isFluentBitConfiguration(source: string) {
    return isFluentBit(source);
  }
  get AST(): FluentBitSchemaType[] {
    return tokensToAST(this._tokens);
  }

  get filePath() {
    return this._filePath;
  }
  get source() {
    return this._source;
  }
  get schema() {
    return getFluentBitSchema(this.AST);
  }
  toString(indent?: number) {
    return schemaToString(this.schema, { propIndent: indent });
  }
}
