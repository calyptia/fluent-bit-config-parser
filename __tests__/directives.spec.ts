import { readFileSync } from 'fs';
import { FluentBitSchema } from '../index';
import { TokenError } from '../src/TokenError';

jest.mock('uuid', () => ({ v4: () => 'UNIQUE' }));
describe('Fluent Bit: Directives', () => {
  it('Not recognized directives should fail', () => {
    const filePath = '/__fixtures__/directives/directives/ephemeral.conf';
    const rawConfig = `
    @SET A=B
    @WHATEVER something that we do not have implemented/invalid. 

    [INPUT]
      name dummy
      dummy {"message":"\${A}"}

    `;

    try {
      new FluentBitSchema(rawConfig, filePath);
    } catch (e) {
      const error = e as TokenError;
      expect(error.message).toMatchInlineSnapshot(
        '"You have defined a directive that cannot be parse (@WHATEVER something that we do not have implemented/invalid. ). The supported directives are: SET,INCLUDE"'
      );
      expect(error.formattedError).toMatchInlineSnapshot(
        '"/__fixtures__/directives/directives/ephemeral.conf: 3:5 You have defined a directive that cannot be parse (@WHATEVER something that we do not have implemented/invalid. ). The supported directives are: SET,INCLUDE"'
      );
    }
    expect.hasAssertions();
  });
  describe('@INCLUDE', () => {
    it('Parses global @INCLUDES in configuration (ignoreFullPath: false)', () => {
      const filePath = '/__fixtures__/directives/include/ephemeral';
      const rawConfig = `
      @INCLUDE /this/is/a/full/path.conf

      [INPUT]
        name dummy
      `;
      try {
        new FluentBitSchema(rawConfig, filePath);
      } catch (e) {
        const error = e as TokenError;

        expect(error.filePath).toMatchInlineSnapshot('"/__fixtures__/directives/include/ephemeral"');
        expect(error.message).toMatchInlineSnapshot('"Can not find file /this/is/a/full/path.conf"');
      }
      expect.hasAssertions();
    });
    it('Parses global @INCLUDES in configuration (ignoreFullPath: true)', () => {
      const filePath = '/__fixtures__/directives/include/ephemeral';
      const rawConfig = `
      @INCLUDE /this/is/a/full/path.conf

      [INPUT]
        name dummy
      `;
      const config = new FluentBitSchema(rawConfig, filePath, { ignoreFullPaths: true });
      expect(config.directives).toMatchObject([]);
      expect(config.AST).toMatchInlineSnapshot(`
        Array [
          Object {
            "__filePath": "/__fixtures__/directives/include/ephemeral",
            "command": "INPUT",
            "id": "UNIQUE",
            "name": "dummy",
            "optional": Object {},
          },
        ]
      `);
    });
    it('Parses global @INCLUDES in configuration', () => {
      const filePath = './__fixtures__/directives/include/withIncludes.conf';
      const rawConfig = readFileSync(filePath, { encoding: 'utf-8' });
      const config = new FluentBitSchema(rawConfig, filePath);
      expect(config.directives).toMatchInlineSnapshot(`
        Array [
          Object {
            "col": 1,
            "filePath": "<PROJECT_ROOT>/__fixtures__/directives/include/nested/tail.conf",
            "line": 6,
            "lineBreaks": 0,
            "offset": 47,
            "text": "@include nested/tail.conf",
            "toString": [Function],
            "type": "INCLUDE",
            "value": "@INCLUDE nested/tail.conf",
          },
          Object {
            "col": 1,
            "filePath": "<PROJECT_ROOT>/__fixtures__/directives/include/nested/service.conf",
            "line": 8,
            "lineBreaks": 0,
            "offset": 136,
            "text": "@InClUdE nested/service.conf",
            "toString": [Function],
            "type": "INCLUDE",
            "value": "@INCLUDE nested/service.conf",
          },
        ]
      `);
      expect(config.AST).toMatchInlineSnapshot(`
              Array [
                Object {
                  "__filePath": "<PROJECT_ROOT>/__fixtures__/directives/include/nested/tail.conf",
                  "command": "INPUT",
                  "id": "UNIQUE",
                  "name": "tail",
                  "optional": Object {
                    "alias": "function_A_json_tail",
                    "parser": "json",
                    "path": "\${DEFAULT_LOGS_DIR}/some-json.log",
                    "path_key": "filename",
                    "refresh_interval": "10",
                    "skip_empty_lines": "On",
                    "skip_long_lines": "On",
                    "tag": "recommended.log.functionA",
                  },
                },
                Object {
                  "__filePath": "<PROJECT_ROOT>/__fixtures__/directives/include/nested/service.conf",
                  "command": "SERVICE",
                  "id": "UNIQUE",
                  "optional": Object {
                    "flush": "1",
                    "health_check": "On",
                    "http_port": "\${HTTP_PORT}",
                    "http_server": "On",
                    "log_level": "Debug",
                    "parsers_file": "/fluent-bit/etc/parsers/parsers-custom.conf",
                    "storage.metrics": "On",
                  },
                },
                Object {
                  "__filePath": "<PROJECT_ROOT>/__fixtures__/directives/include/withIncludes.conf",
                  "command": "OUTPUT",
                  "id": "UNIQUE",
                  "name": "loki",
                  "optional": Object {
                    "alias": "loki_output",
                    "host": "\${LOKI_HOST}",
                    "label_keys": "$file,$level",
                    "labels": "job=recommended-fluentbit",
                    "match": "\${LOKI_MATCH}",
                    "port": "\${LOKI_PORT}",
                    "workers": "1",
                  },
                },
              ]
          `);
    });
    it('Fails retrieving an @INCLUDE that contains more than a single path as a value', async () => {
      const filePath = '__fixtures__/directives/include/withWrongIncludeValue.conf';
      const rawConfig = readFileSync(filePath, { encoding: 'utf-8' });
      try {
        new FluentBitSchema(rawConfig, filePath);
      } catch (e) {
        expect(e).toBeInstanceOf(TokenError);
        const error = e as TokenError;
        expect(error.line).toBe(3);
        expect(error.col).toBe(1);
        expect(error.message).toMatchInlineSnapshot(
          '"You are trying to include nested/tail.conf, but we also found more arguments (shouldNotHaveAnytingElse). @INCLUDE directive can only have a single value (ex: @INCLUDE path/to/a/file)"'
        );
        expect(error.formattedError).toMatchInlineSnapshot(
          '"<PROJECT_ROOT>/__fixtures__/directives/include/withWrongIncludeValue.conf: 3:1 You are trying to include nested/tail.conf, but we also found more arguments (shouldNotHaveAnytingElse). @INCLUDE directive can only have a single value (ex: @INCLUDE path/to/a/file)"'
        );
        expect(error.filePath).toMatchInlineSnapshot(
          '"<PROJECT_ROOT>/__fixtures__/directives/include/withWrongIncludeValue.conf"'
        );
      }
      expect.hasAssertions();
    });
    it('Fails retrieving a repeated @INCLUDE (can not include file twice)', async () => {
      const filePath = '__fixtures__/directives/include/withDuplicatedIncludes.conf';
      const rawConfig = readFileSync(filePath, { encoding: 'utf-8' });
      try {
        new FluentBitSchema(rawConfig, filePath);
      } catch (e) {
        expect(e).toBeInstanceOf(TokenError);
        const error = e as TokenError;
        expect(error.line).toBe(9);
        expect(error.col).toBe(1);
        expect(error.message).toMatchInlineSnapshot(
          '"You are trying to include <PROJECT_ROOT>/__fixtures__/directives/include/nested/tail.conf. Fluent Bit does not allow a file to be included twice in the same configuration"'
        );
        expect(error.formattedError).toMatchInlineSnapshot(
          '"<PROJECT_ROOT>/__fixtures__/directives/include/withDuplicatedIncludes.conf: 9:1 You are trying to include <PROJECT_ROOT>/__fixtures__/directives/include/nested/tail.conf. Fluent Bit does not allow a file to be included twice in the same configuration"'
        );
        expect(error.filePath).toMatchInlineSnapshot(
          '"<PROJECT_ROOT>/__fixtures__/directives/include/withDuplicatedIncludes.conf"'
        );
      }
      expect.hasAssertions();
    });
    it('Fails retrieving a missing @INCLUDE (file not found) ', async () => {
      const filePath = './__fixtures__/directives/include/withFailingIncludes.conf';
      const rawConfig = readFileSync(filePath, { encoding: 'utf-8' });
      try {
        new FluentBitSchema(rawConfig, filePath);
      } catch (e) {
        expect(e).toBeInstanceOf(TokenError);
        const error = e as TokenError;
        expect(error.line).toBe(3);
        expect(error.col).toBe(1);
        expect(error.message).toMatchInlineSnapshot('"Can not find file nested/notExistentInclude.conf"');
        expect(error.formattedError).toMatchInlineSnapshot(
          '"<PROJECT_ROOT>/__fixtures__/directives/include/withFailingIncludes.conf: 3:1 Can not find file nested/notExistentInclude.conf"'
        );
        expect(error.filePath).toMatchInlineSnapshot(
          '"<PROJECT_ROOT>/__fixtures__/directives/include/withFailingIncludes.conf"'
        );
      }
      expect.hasAssertions();
    });
  });
  describe('@SET', () => {
    it('Parses @SET in configuration', () => {
      const filePath = './__fixtures__/directives/set/withSetVars.conf';
      const rawConfig = readFileSync(filePath, { encoding: 'utf-8' });
      const config = new FluentBitSchema(rawConfig, filePath);
      expect(config.directives).toMatchInlineSnapshot(`
        Array [
          Object {
            "col": 1,
            "filePath": "<PROJECT_ROOT>/__fixtures__/directives/set/withSetVars.conf",
            "line": 1,
            "lineBreaks": 0,
            "offset": 0,
            "text": "@SET A=some configuration here again",
            "toString": [Function],
            "type": "SET",
            "value": "@SET A=some configuration here again",
          },
          Object {
            "col": 1,
            "filePath": "<PROJECT_ROOT>/__fixtures__/directives/set/withSetVars.conf",
            "line": 2,
            "lineBreaks": 0,
            "offset": 37,
            "text": "@set C=some configuration here",
            "toString": [Function],
            "type": "SET",
            "value": "@SET C=some configuration here",
          },
        ]
      `);
      expect(config.AST).toMatchInlineSnapshot(`
        Array [
          Object {
            "__filePath": "<PROJECT_ROOT>/__fixtures__/directives/set/withSetVars.conf",
            "command": "INPUT",
            "id": "UNIQUE",
            "name": "dummy",
            "optional": Object {
              "dummy": "{\\"message\\":\\"\${A}\\"}",
            },
          },
          Object {
            "__filePath": "<PROJECT_ROOT>/__fixtures__/directives/set/withSetVars.conf",
            "command": "INPUT",
            "id": "UNIQUE",
            "name": "dummy",
            "optional": Object {
              "dummy": "{\\"message\\":\\"\${C}\\"}",
            },
          },
          Object {
            "__filePath": "<PROJECT_ROOT>/__fixtures__/directives/set/withSetVars.conf",
            "command": "OUTPUT",
            "id": "UNIQUE",
            "name": "stdout",
            "optional": Object {
              "match": "*",
            },
          },
        ]
      `);
    });

    it('Should parse the @SET directive correctly', () => {
      const filePath = '/__fixtures__/directives/set/ephemeral.conf';
      const rawConfig = `
      @SET A = some configuration here again =
      @set C=some configuration here
      
      [INPUT]
        name dummy
        dummy {"message":"\${A}"}
      
      `;

      const config = new FluentBitSchema(rawConfig, filePath);

      expect(config.directives).toMatchInlineSnapshot(`
        Array [
          Object {
            "col": 7,
            "filePath": "/__fixtures__/directives/set/ephemeral.conf",
            "line": 2,
            "lineBreaks": 0,
            "offset": 7,
            "text": "@SET A = some configuration here again =",
            "toString": [Function],
            "type": "SET",
            "value": "@SET A = some configuration here again =",
          },
          Object {
            "col": 7,
            "filePath": "/__fixtures__/directives/set/ephemeral.conf",
            "line": 3,
            "lineBreaks": 0,
            "offset": 54,
            "text": "@set C=some configuration here",
            "toString": [Function],
            "type": "SET",
            "value": "@SET C=some configuration here",
          },
        ]
      `);
    });

    it('Parses @SET in configuration (case insensitive)', () => {
      const filePath = '/__fixtures__/directives/set/ephemeral.conf';
      const rawConfig = `
      @Set A=A
      @SeT B=B
      @SEt C=C
      @seT D=D
      @sET E=E
      @sEt F=F
      
      [INPUT]
        name dummy
        dummy {"message":"\${A}"}
      `;
      const config = new FluentBitSchema(rawConfig, filePath);
      expect(config.directives).toMatchSnapshot();
    });
    it('Should parse the @SET directive correctly, even when there is a space after the assignment', () => {
      const filePath = '/__fixtures__/directives/set/ephemeral.conf';
      const rawConfig = `
      # Note the space which becomes part of the variable name, we should not do this. but is allowed :/
      @set C =3
      
      [INPUT]
        name dummy
        dummy {"message":"\${C }"}
      
      `;
      const config = new FluentBitSchema(rawConfig, filePath);

      expect(config.directives).toMatchInlineSnapshot(`
        Array [
          Object {
            "col": 7,
            "filePath": "/__fixtures__/directives/set/ephemeral.conf",
            "line": 3,
            "lineBreaks": 0,
            "offset": 112,
            "text": "@set C =3",
            "toString": [Function],
            "type": "SET",
            "value": "@SET C =3",
          },
        ]
      `);
    });

    it('Should add the @SET directive when calling toString()', () => {
      const filePath = '/__fixtures__/directives/set/ephemeral.conf';
      const rawConfig = `
      @SET A=some configuration here again =
      @set C=some configuration here
      
      [INPUT]
        name dummy
        dummy {"message":"\${A}"}
      `;

      const config = new FluentBitSchema(rawConfig, filePath);
      expect(config.toString()).toMatchInlineSnapshot(`
        "
        @SET A=some configuration here again =

        @SET C=some configuration here
                                   
        [INPUT]                    
          name  dummy              
          dummy {\\"message\\":\\"\${A}\\"} 
        "
      `);
    });
    it('Should fail when @SET directive is malformed', async () => {
      const filePath = './__fixtures__/directives/set/withManySets.conf';
      const rawConfig = readFileSync(filePath, { encoding: 'utf-8' });
      try {
        new FluentBitSchema(rawConfig, filePath);
      } catch (e) {
        expect(e).toBeInstanceOf(TokenError);
        const error = e as TokenError;
        expect(error.line).toBe(5);
        expect(error.col).toBe(1);
        expect(error.message).toMatchInlineSnapshot(
          '"You have defined a directive that cannot be parse (@set ${A}=1). The supported directives are: SET,INCLUDE"'
        );
        expect(error.formattedError).toMatchInlineSnapshot(
          '"<PROJECT_ROOT>/__fixtures__/directives/set/withManySets.conf: 5:1 You have defined a directive that cannot be parse (@set ${A}=1). The supported directives are: SET,INCLUDE"'
        );
        expect(error.filePath).toMatchInlineSnapshot('"<PROJECT_ROOT>/__fixtures__/directives/set/withManySets.conf"');
      }
      expect.hasAssertions();
    });
  });
});
