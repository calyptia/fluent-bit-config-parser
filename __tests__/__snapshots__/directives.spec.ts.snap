// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fluent Bit: Directives @SET Parses @SET in configuration (case insensitive) 1`] = `
Array [
  Object {
    "col": 7,
    "filePath": "/__fixtures__/directives/set/ephemeral.conf",
    "line": 2,
    "lineBreaks": 0,
    "offset": 7,
    "text": "@Set A=A",
    "toString": [Function],
    "type": "SET",
    "value": "@SET A=A",
  },
  Object {
    "col": 7,
    "filePath": "/__fixtures__/directives/set/ephemeral.conf",
    "line": 3,
    "lineBreaks": 0,
    "offset": 22,
    "text": "@SeT B=B",
    "toString": [Function],
    "type": "SET",
    "value": "@SET B=B",
  },
  Object {
    "col": 7,
    "filePath": "/__fixtures__/directives/set/ephemeral.conf",
    "line": 4,
    "lineBreaks": 0,
    "offset": 37,
    "text": "@SEt C=C",
    "toString": [Function],
    "type": "SET",
    "value": "@SET C=C",
  },
  Object {
    "col": 7,
    "filePath": "/__fixtures__/directives/set/ephemeral.conf",
    "line": 5,
    "lineBreaks": 0,
    "offset": 52,
    "text": "@seT D=D",
    "toString": [Function],
    "type": "SET",
    "value": "@SET D=D",
  },
  Object {
    "col": 7,
    "filePath": "/__fixtures__/directives/set/ephemeral.conf",
    "line": 6,
    "lineBreaks": 0,
    "offset": 67,
    "text": "@sET E=E",
    "toString": [Function],
    "type": "SET",
    "value": "@SET E=E",
  },
  Object {
    "col": 7,
    "filePath": "/__fixtures__/directives/set/ephemeral.conf",
    "line": 7,
    "lineBreaks": 0,
    "offset": 82,
    "text": "@sEt F=F",
    "toString": [Function],
    "type": "SET",
    "value": "@SET F=F",
  },
]
`;
